flink-metrics/flink-metrics-datadog/src/main/java/org/apache/flink/metrics/datadog/DatadogHttpReporter.java

```diff
@@ -157,6 +181,11 @@ private void addGaugesAndUnregisterOnException(DSeries request) {
 				// Flink uses Gauge to store many types other than Number
 				g.getMetricValue();
 				request.addGauge(g);
+				++currentCount;
+				if (currentCount % maxMetricsPerRequestValue == 0 || currentCount >= totalGauges) {
+					client.send(request);
```

Comment:
hmm, I don't really like that we're always sending at least 2 requests, as long as at least one counter and gauge exist.

Ultimately, all we need to do is slice the currently created `DSeries`, which we could do using `ArrayList#subList`.

```
int fromIndex = 0;
ArrayList<DMetric> series = request.getSeries();
while (fromIndex < series.size()) {
  int toIndex = Math.Min(currentCount + maxMetricsPerRequestValue, series.size());
  client.send(series.subList(fromIndex, toIndex));
  fromIndex = toIndex;
}
```

WDYT? We could even wrap this logic into a stateful `Consumer` that resets the `DSeries` on each `client.send`, so we don't have to assemble the full list.

Reply:
I do like the sublist approach although `DatadogHttpClient` serializes a `DSeries` not a `List`.

maybe something like:

```
List<DMetric> metrics = new ArrayList();
addGaugesAndUnregisterOnException(metrics);
metrics.addAll(counters.values());
metrics.addAll(meters.values());

int fromIndex = 0;
while (fromIndex < metrics.size()) {
    int toIndex = Math.Min(fromIndex + maxMetricsPerRequestValue, metrics.size());
    client.send(new DSeries(metrics.subList(fromIndex, toIndex)));
    fromIndex = toIndex;
}
```

Reply:
yes, you'D of course have to rap the subList in a DSeries again, requiring a new constructor.